pkg/handlers/v1/attribute_test.go

```diff
@@ -23,11 +28,10 @@ func TestSuccess(t *testing.T) {
 
 	handler := &AttributeHandler{
 		AssetAttributor: assetattributor.NewNoOpAssetAttributor(),
-		LogFn:           domain.LoggerFromContext,
+		LogFn:           newNoOpLogger,
```

Comment:
I think I am personally still unclear about the advantages of using mocks over using Nop implementations in this specific case. I personally prefer using (and have done) what Loren is proposing.

To me, anytime you use a mock, you are needing to do some kind of behavior or state verification. Anytime the interface or behavior changes, changes to the test are required. In this case, rather than a mock, we are using a fake logger since log calls aren't really something we need to track or verify. Additionally if we need to add/remove logs from production code, I don't think that should necessitate a change in a test file.

Reply:
I understand and agree with what you're saying @zlozano . I think my use of the term "mock" is more general than yours. Whether we achieve on the goal with a mock, fixture, or stub, etc. is not as important to me. What I'm calling out is that this version of the tests doesn't use a "fake logger" or a "fake stat" because it relies on specific details of the implementations which are supposed to be hidden. They are the "real" versions we'd get at runtime which means their internal behavior is unknown and may have any number of side-effects. The fact that we aren't getting bunch of log spam output to `stdout` is only hiding this fact.

re: interfaces. Again, I entirely agree. If the interfaces change then a refactor is required. However, the specific side-effects of any given implementation are not part of the contract provided by the interfaces. It is specifically the managing of these implementation specific side-effects that I'm calling out as a problem. For example, there is no "contractual" obligation within the current major version of `xstats` that would prevent it from swapping the `NOP` client for a `STDOUT` client in cases where the context is empty. Alternatively, we could move away from `xstats` to a new implementation that panics when the context is empty similar to how `logevent` handles empty contexts. The bottom line is that we should not rely on these behaviors for our tests to function correctly.

My overall statement here is: "Log and stat clients are _no different from any other dependency in our code_ and should not be treated differently."

Reply:
Ah that makes sense. Your xstats example made things clear. 👍

Reply:
I recently went through this same exercise in another one of our projects. Here's a log and stat stub:

```golang
type nopLogger struct{}

func (*nopLogger) Debug(event interface{})                 {}
func (*nopLogger) Info(event interface{})                  {}
func (*nopLogger) Warn(event interface{})                  {}
func (*nopLogger) Error(event interface{})                 {}
func (*nopLogger) SetField(name string, value interface{}) {}
func (logger *nopLogger) Copy() Logger {
    return logger
}

var testLogger = &nopLogger{}

func testLogFn(context.Context) Logger { return testLogger }

type nopStat struct{}

func (*nopStat) Gauge(stat string, value float64, tags ...string)        {}
func (*nopStat) Count(stat string, count float64, tags ...string)        {}
func (*nopStat) Histogram(stat string, value float64, tags ...string)    {}
func (*nopStat) Timing(stat string, value time.Duration, tags ...string) {}
func (*nopStat) AddTags(tags ...string)                                  {}
func (*nopStat) GetTags() []string {
    return []string{}
}

var testStat = &nopStat{}

func testStatFn(context.Context) Stat { return testStat }
```